ktor-network/ktor-network-quic/jvmAndNix/src/io/ktor/network/quic/frames/FrameWriter.kt

```diff
@@ -0,0 +1,544 @@
+        sequenceNumber: Long,
+    ) = with(packetBuilder) {
+        writeFrameType(FrameType_v1.RETIRE_CONNECTION_ID)
+        writeVarInt(sequenceNumber)
+    }
+
+    override fun writePathChallenge(
+        packetBuilder: BytePacketBuilder,
+        data: ByteArray,
+    ) = with(packetBuilder) {
+        require(data.size == 8) {
+            "The size of the value in the 'Data' field in PATH_CHALLENGE frame MUST be 8 bytes"
+        }
+        writeFrameType(FrameType_v1.PATH_CHALLENGE)
+        writeFully(data)
+    }
+
+    override fun writePathResponse(
+        packetBuilder: BytePacketBuilder,
+        data: ByteArray,
+    ) = with(packetBuilder) {
+        require(data.size == 8) {
+            "The size of the value in the 'Data' field in PATH_RESPONSE frame MUST be 8 bytes"
+        }
+        writeFrameType(FrameType_v1.PATH_RESPONSE)
+        writeFully(data)
+    }
+
+    override fun writeConnectionCloseWithTransportError(
+        packetBuilder: BytePacketBuilder,
+        errorCode: QUICTransportError_v1,
+        frameTypeV1: FrameType_v1?,
+        reasonPhrase: ByteArray,
+    ) = writeConnectionClose(
+        packetBuilder = packetBuilder,
+        typeV1 = FrameType_v1.CONNECTION_CLOSE_TRANSPORT_ERR,
+        errorCode = errorCode,
+        frameTypeV1 = frameTypeV1 ?: FrameType_v1.PADDING,
+        reasonPhrase = reasonPhrase
+    )
+
+    override fun writeConnectionCloseWithAppError(
+        packetBuilder: BytePacketBuilder,
+        errorCode: AppError_v1,
+        reasonPhrase: ByteArray,
+    ) = writeConnectionClose(
+        packetBuilder = packetBuilder,
+        typeV1 = FrameType_v1.CONNECTION_CLOSE_APP_ERR,
+        errorCode = errorCode,
+        frameTypeV1 = null,
+        reasonPhrase = reasonPhrase
+    )
+
+    private fun writeConnectionClose(
+        packetBuilder: BytePacketBuilder,
+        typeV1: FrameType_v1,
+        errorCode: Error_v1,
+        frameTypeV1: FrameType_v1?,
+        reasonPhrase: ByteArray,
+    ) = with(packetBuilder) {
+        writeFrameType(typeV1)
+        writeErrorCode(errorCode)
+        frameTypeV1?.let { writeFrameType(it) }
+        writeVarInt(reasonPhrase.size)
+        writeFully(reasonPhrase)
+    }
+
+    override fun writeHandshakeDone(
+        packetBuilder: BytePacketBuilder,
+    ) = with(packetBuilder) {
+        writeFrameType(FrameType_v1.HANDSHAKE_DONE)
+    }
+
+    // HELPER FUNCTIONS AND VALUES
+
+    @OptIn(ExperimentalUnsignedTypes::class)
+    private fun BytePacketBuilder.writeFrameType(typeV1: FrameType_v1) {
+        // it is actually a varint with length 8, as frame types are all values in 0x00..0x1e
+        writeUByte(typeV1.typeValue)
+    }
+
+    @OptIn(ExperimentalUnsignedTypes::class)
+    private fun BytePacketBuilder.writeConnectionId(id: ByteArray) {
+        writeUByte(id.size.toUByte())
+        writeFully(id)
+    }
+
+    private fun BytePacketBuilder.writeErrorCode(errorCode: Error_v1) {
```

Comment:
BTW, this will also involve boxing for inherited `inline` classes. May be just making this function `inline` will allow to not box it, but Im not sure

Reply:
I decided to get rid of Error_v1 interface at all, as it is the only place it is used, and app and transport error are processed separately